src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,481 @@
+    if (shape[shape.length() - 1] == '\'')
+        shape = shape.substr(0, shape.length() - 1);
+
+    NumpyHeader res;
+    res.shape = parseShape(shape);
+    res.numpy_type = parseType(descr);
+
+    return res;
+}
+
+DataTypePtr getNestedType(DataTypePtr type)
+{
+    String a = type->getName();
+    while (const auto * temp_type = typeid_cast<const DataTypeArray *>(type.get()))
+    {
+        type = temp_type->getNestedType();
+        a = type->getName();
+    }
+
+    return type;
+}
+}
+
+void NpyRowInputFormat::readPrefix()
+{
+    header = parseHeader(*in);
+}
+
+NpyRowInputFormat::NpyRowInputFormat(ReadBuffer & in_, Block header_, Params params_)
+    : IRowInputFormat(std::move(header_), in_, std::move(params_))
+{
+    auto types = getPort().getHeader().getDataTypes();
+    if (types.size() != 1)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unexpected number of columns for Npy input format, expected one column, got {} columns", types.size());
+    nested_type = getNestedType(types[0]);
+}
+
+void NpyRowInputFormat::readRows(MutableColumns & columns)
+{
+    auto & column = columns[0];
+    IColumn * current_column = column.get();
+    size_t elements_in_current_column = 1;
+    for (size_t i = 1; i != header.shape.size(); ++i)
+    {
+        auto & array_column = assert_cast<ColumnArray &>(*current_column);
+        /// Fill offsets of array columns.
+        for (size_t j = 0; j != elements_in_current_column; ++j)
+            array_column.getOffsets().push_back(array_column.getOffsets().back() + header.shape[i]);
+        current_column = &array_column.getData();
+        elements_in_current_column *= header.shape[i];
+    }
+
+    for (size_t i = 0; i != elements_in_current_column; ++i)
+        chooseType(current_column);
+}
+
+template <typename ColumnValue, typename DataValue>
+void NpyRowInputFormat::readBinaryValueAndInsert(MutableColumnPtr column, NumpyDataType::Endianness endianess)
+{
+    DataValue value;
+    if (endianess == NumpyDataType::Endianness::BIG)
+        readBinaryBigEndian(value, *in);
+    else
+        readBinaryLittleEndian(value, *in);
+    assert_cast<ColumnVector<ColumnValue> &>(*column).insertValue(static_cast<ColumnValue>(value));
+}
+
+void NpyRowInputFormat::readStringBinaryAndInsert(MutableColumnPtr column, size_t size, bool is_fixed)
+{
+    if (is_fixed)
+    {
+        auto & fixed_string_column = assert_cast<ColumnFixedString &>(*column);
+        size_t n = fixed_string_column.getN();
+        if (size > n)
+            throw Exception(ErrorCodes::TOO_LARGE_STRING_SIZE, "Too large string for FixedString column");
+        auto & data = fixed_string_column.getChars();
+
+        size_t old_size = data.size();
+        data.resize_fill(old_size + n);
+
+        String tmp;
+        tmp.resize(size);
+
+        in->readStrict(tmp.data(), size);
+        tmp.erase(std::remove(tmp.begin(), tmp.end(), '\0'), tmp.end());
+        fixed_string_column.insertData(tmp.c_str(), tmp.size());
```

Comment:
I think you mixed up 2 ways of inserting data into FixedString column, directly to `chars` and using `insertData`. Let's insert directly to chars:
```suggestion
        size_t old_size = data.size();
        data.resize_fill(old_size + n);
        in->readStrict(data + old_size, size);
```